libr/parse/p/parse_x86_pseudo.c

```diff
@@ -358,12 +364,58 @@ static inline void mk_reg_str(const char *regname, int delta, bool sign, bool at
 	r_strbuf_free (sb);
 }
 
-static bool varsub (RParse *p, RAnalFunction *f, ut64 addr, int oplen, char *data, char *str, int len) {
-	RList *bpargs, *spargs;
+static const RList *__args(RParse *p, RAnalFunction *f, int type) {
+#if CACHE_VARS
+	static RAnalFunction *old_f = NULL;
+	static RList *__cache_spargs_list = NULL;
+	static RList *__cache_bpargs_list = NULL;
+	if (type == -1 && !p && !f) {
+		// free to issue mamleak
+		r_list_free (__cache_bpargs_list);
+		r_list_free (__cache_spargs_list);
+		old_f = NULL;
+		return NULL;
+	}
+	r_return_val_if_fail (p, NULL);
+	if (!f || !f->anal) {
```

Comment:
do a `r_return_val_if_fail (f && f->anal, NULL)`. If you are calling `__args` without a function or with broken anal, you have already screwed things up a while ago.

Reply:
i would prefer to focus on the cache in this PR, this code have a lot of things to improve

Reply:
I *think* the cache method is ok ebough, but you should tell us what the benchmark says.